library/cartridge_apply_app_config.py

```diff
@@ -0,0 +1,231 @@
+    if tdg_version[0] == 'scm-1':
+        return {
+            'auth-token': tdg_token,  # TDG <= 1.6
+            'Authorization': 'Bearer ' + tdg_token,  # TDG >= 1.7
+        }
+
+    major = int(tdg_version[0])
+    minor = int(tdg_version[1])
+    if major < 1 or major == 1 and minor <= 6:
+        return {'auth-token': tdg_token}
+    return {'Authorization': 'Bearer ' + tdg_token}
+
+
+def get_upload_endpoint(upload_endpoint, remote_config_path):
+    if upload_endpoint is not None:
+        upload_endpoint = upload_endpoint.lower()
+
+    if not os.path.isdir(remote_config_path):
+        if os.path.splitext(remote_config_path)[1] in ['.yml', '.yaml']:
+            upload_endpoint = upload_endpoint or LUA_ENDPOINT
+        else:
+            upload_endpoint = upload_endpoint or HTTP_ENDPOINT
+
+        assert upload_endpoint in [LUA_ENDPOINT, HTTP_ENDPOINT], \
+            'File config available only for Lua or HTTP endpoints'
+    else:
+        upload_endpoint = upload_endpoint or HTTP_ENDPOINT
+
+        assert upload_endpoint in [HTTP_ENDPOINT, TDG_ENDPOINT], \
+            'Directory config available only for HTTP or TDG endpoints'
+
+    return upload_endpoint
+
+
+def get_apply_config_func(upload_endpoint, console_sock=None, http_port=None, cluster_cookie=None, tdg_token=None):
+    dir_mode = ZIP_DIR_MODE
+
+    if upload_endpoint == LUA_ENDPOINT:
+        assert console_sock is not None, "Console socket is required for Cartridge endpoint"
+        control_console = helpers.get_control_console(console_sock)
+
+        def apply_config(path):
+            return patch_file_clusterwide(control_console, path)
+
+    elif upload_endpoint == HTTP_ENDPOINT:
+        assert cluster_cookie is not None, 'Cluster cookie is required for HTTP endpoint'
+        assert http_port is not None, 'HTTP port is required for HTTP endpoint'
+        headers = {'Authorization': basic_auth_header('admin', cluster_cookie)}
+
+        def apply_config(path):
+            return send_on_http(http_port, headers, path)
+
+    elif upload_endpoint == TDG_ENDPOINT:
+        tdg_upload_mode = get_tdg_upload_mode(console_sock)
+        if tdg_upload_mode == 'http':
+            assert http_port is not None, 'HTTP port is required for TDG endpoint'
+            headers = get_tdg_auth_headers(console_sock, tdg_token)
+
+            def apply_config(path):
+                return send_on_http(http_port, headers, path)
+
+        elif tdg_upload_mode == 'lua':
+            dir_mode = PURE_DIR_MODE
+
+            def apply_config(path):
+                return apply_tdg_config(console_sock, path)
+
+        else:
+            raise AssertionError("Unknown TDG upload mode '%s'" % tdg_upload_mode)
+
+    else:
+        raise AssertionError("Unknown upload endpoint '%s'" % upload_endpoint)
+
+    return apply_config, dir_mode
+
+
+def apply_app_config(params):
+    remote_config_path = os.path.join(params['remote_dir'], os.path.basename(params['local_config_path']))
```

Comment:
I would like `remote_dir` to be named `tmp_dir`.

Reply:
But this is the directory where the config is located, it is not necessarily temporary